### ISSUE TITLE
Unable to load audio files in jar file

### ISSUE THREAD

**Issue by fgnm** [rednblackgames/gdx-miniaudio#2]
Currently, `gdx-miniaudio` cannot read audio files placed inside the `.jar` archive. There are a couple way to fix this on the library side, but any fix looks out of the scope of this library. At moment the most efficient way to load audio files packed in a distribution `jar` is to extract them to a temporary directory and load files with absolute/relative path.

**Comment by Geolykt** [comment]
Why isn't the usage of decoders a possibility? It would allow the usage of `ma_decoder_init_memory`, which would ultimately mean that the PCM data from FLACs, MP3s or WAVs could easily be gathered without depending on other libraries. Of course, doing so probably would mean creating an MADecoder class.

From the rough MA calls needed you'd probably have something similar to (of course grossly over-exaggerated)

```c
ma_decoder decoder;
ma_decoder_init_memory(data, length, NULL /* (The decoder config) */, &decoder);
ma_result result = ma_decoder_read_pcm_frames(&decoder, framesOut, frameCount, &framesOutCount);
ma_decoder_uninit(&decoder);
```

**Comment by fgnm** [comment]
Yes you're right, a MADecoder binding would be useful in this and also other cases.. I'm also sorry that the binding is stuck at `v0.11.14` while MiniAudio is now at `v0.11.21`, unfortunately I lost my way to test macOS and iOS.. But I know, I need to keep this updated in a way or another, I'll try to find a solution soon :)

**Comment by Geolykt** [comment]
I unfortunately cannot help on that front

**Comment by tinspin** [comment]
Is this still a problem.

You cannot use royalty free sounds if you extract them on the disk.

**Comment by fgnm** [comment]
Sorry my late response! Yes, in v0.5 the issue is partially fixed because you can r3ad bytes from a FileHandle and decode them with `MiniAudio#decodeBytes`, however I don't consider this as a full solution. The best one is to allow native C code to read directly inside the jar file using another custom ResourceManager, most likely based on something like this: https://github.com/mackron/fs

I'm still looking for a simple but effective library for the job.. if someone has suggestions please tell me :)

**Comment by nulldoubt** [comment]
Hey, how about copying the file to a cached location then referring to it externally in the native code?

I've seen that approach in the Arc fork of libGDX, [here](https://github.com/Anuken/Arc/blob/master/arc-core/src/arc/audio/Music.java#L47), but they use Soloud instead of MiniAudio.

**Comment by fgnm** [comment]
Well caching is however the extraction from the jar to a temporary/persistent location, which at moment is the most effective solution. But it's something I would still leave to user code and not include in the library itself

**Comment by SomeTroglodyte** [comment]
Clarifications:
- The solution commit is a5dfe8548a62d8a134a421551d360516d7b694ff - it wasn't autolinked, the closing keyword in the commit message closed issue 1 instead.
- As stated, implemented in MASoundLoader. That means if you use `MiniAudio.createSound`, you still need your own workaround (catch-readBytes-decodeBytes).
- If you decide to use `MASoundLoader.loadAsync`, note that requires a Gdx `FileHandleResolver` and an Android `AssetManager`, and waiting for `loadSync` to return non-null.
- Also note the new `loadAsync` will only fix the issue when you pass null for the `MASoundLoaderParameters parameter`. With parameter, it behaves as before and will bypass the issue only if `loadFromMemory` is true.
